phi/tools/firecrawl.py

```diff
@@ -44,33 +62,145 @@ def scrape_website(self, url: str) -> str:
+            raise ValueError("No URL provided. URL is required for crawling.")
 
-        params: Dict[str, Any] = {}
+        params = {}
         if self.limit or limit:
             params["limit"] = self.limit or limit
-            if self.formats:
-                params["scrapeOptions"] = {"formats": self.formats}
+            if self.formats and set(self.formats).issubset(set(["markdown", "html", "rawHtml", "links", "screenshot"])):
+                params["scrapeOptions"] = {"formats": self.formats}  # type: ignore
 
-        crawl_result = self.app.crawl_url(url, params=params, poll_interval=30)
+        crawl_result = self.app.crawl_url(url, params=params, poll_interval=15)
         return json.dumps(crawl_result)
+
+    def async_crawl_website(self, url: str, limit: Optional[int]) -> str:
+        """Use this function to Asynchronously crawl a website using Firecrawl.
+
+        Args:
+            url (str): The URL to crawl.
+            limit (int): The maximum number of pages to crawl.
+
+        Returns:
+            str: The crawl results in JSON format.
+        """
+        if url is None:
+            raise ValueError("No URL provided. URL is required for crawling.")
+
+        params = {}
+        logger.info("`limit` parameter will be deprecated in future release with `crawl_limit` parameter.")
+        if self.limit or limit:
+            params["limit"] = self.limit or limit
+            if self.formats and set(self.formats).issubset(set(["markdown", "html", "rawHtml", "links", "screenshot"])):
+                params["scrapeOptions"] = {"formats": self.formats} # type: ignore
+        poll_interval: int = 15
+
+        # Firecrawl async_crawl_url returns crawl job id, hence adding custom implementation  with helper functions to return final  async crawl response.
+        try:
+            initial_status = self.app.async_crawl_url(url, params=params)
+            crawl_id = initial_status.get("id")
+
+            if not crawl_id:
+                raise ValueError("No crawl ID received from async crawl initiation")
+
+            # Monitor the crawl until completion
+            while True:
+                status_response = self.app.check_crawl_status(crawl_id)
+                current_status = status_response.get("status")
+
+                if current_status == "completed":
+                    if "data" in status_response:
+                        return json.dumps(status_response)
+                    else:
+                        raise ValueError("Crawl completed but no data was returned")
+
+                elif current_status in ["active", "paused", "pending", "queued", "waiting", "scraping"]:
+                    time.sleep(max(poll_interval, 2))
+
+                else:
+                    raise ValueError(f"Crawl failed or was stopped. Status: {current_status}")
+
+        except Exception as e:
+            return json.dumps({"error": str(e)})
+
+    def check_crawl_status(self, crawl_id: str) -> str:
```

Comment:
Will this be necessary for the AI if the async crawl still polls until completion?